File: lua/problem_60.lua
local M = {}

function M.can_partition(nums)
    local total = 0
    for _, v in ipairs(nums) do total = total + v end
    if total % 2 ~= 0 then return false end
    local target = total // 2
    local dp = {}
    for i = 0, target do dp[i] = false end
    dp[0] = true
    for _, v in ipairs(nums) do
        for j = target, v, -1 do
            dp[j] = dp[j] or dp[j - v]
        end
    end
    return dp[target]
end

function M.problem60()
    print("Problem 60:")
    assert(    M.can_partition({15,5,20,10,35,15,10}))
    assert(not M.can_partition({15,5,20,10,35}))
    assert(    M.can_partition({1,5,11,5}))
    print("All Tests Passed")
end

return M

File: c/problem_60.c
#include "problem_60.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

int can_partition(int *nums, int n) {
    int total = 0;
    for (int i=0;i<n;i++) total+=nums[i];
    if (total%2!=0) return 0;
    int target = total/2;
    int *dp = calloc(target+1, sizeof(int));
    dp[0] = 1;
    for (int i=0;i<n;i++)
        for (int j=target;j>=nums[i];j--)
            dp[j] = dp[j] || dp[j-nums[i]];
    int result = dp[target];
    free(dp);
    return result;
}

void problem60(void) {
    printf("Problem 60:\n");
    int a[] = {15,5,20,10,35,15,10}; assert( can_partition(a,7));
    int b[] = {15,5,20,10,35};       assert(!can_partition(b,5));
    int c[] = {1,5,11,5};            assert( can_partition(c,4));
    printf("All Tests Passed\n");
}
